src/kanata/mod.rs

```diff
@@ -456,49 +456,21 @@ impl Kanata {
         let ms_elapsed = ns_elapsed_with_rem / NS_IN_MS;
         self.time_remainder = ns_elapsed_with_rem % NS_IN_MS;
 
-        let mut extra_ticks: u16 = 0;
-        for _ in 0..ms_elapsed {
-            self.tick_states()?;
-            if let Some(event) = tick_replay_state(
-                &mut self.dynamic_macro_replay_state,
-                self.dynamic_macro_replay_behaviour,
-            ) {
-                self.layout.bm().event(event.key_event());
-                extra_ticks = extra_ticks.saturating_add(event.delay());
-                log::debug!("dyn macro extra ticks: {extra_ticks}, ms_elapsed: {ms_elapsed}");
-            }
-        }
-
-        if ms_elapsed > 0 {
-            for i in 0..(extra_ticks.saturating_sub(ms_elapsed as u16)) {
-                self.tick_states()?;
-                if tick_replay_state(
-                    &mut self.dynamic_macro_replay_state,
-                    self.dynamic_macro_replay_behaviour,
-                )
-                .is_some()
-                {
-                    log::error!("overshot to next event at iteration #{i}, the code is broken!");
-                    break;
-                }
-            }
-
-            self.last_tick = match ms_elapsed {
-                0..=10 => now,
-                // If too many ms elapsed, probably doing a tight loop of something that's quite
-                // expensive, e.g. click spamming. To avoid a growing ms_elapsed due to trying and
-                // failing to catch up, reset last_tick to the "actual now" instead the "past now"
-                // even though that means ticks will be missed - meaning there will be fewer than
-                // 1000 ticks in 1ms on average. In practice, there will already be fewer than 1000
-                // ticks in 1ms when running expensive operations, this just avoids having tens to
-                // thousands of ticks all happening as soon as the expensive operations end.
-                _ => time::Instant::now(),
-            };
+        self.tick_ms(ms_elapsed)?;
+
+        self.last_tick = match ms_elapsed {
+            0..=10 => now,
```

Comment:
Moving this out of `if ms_elapsed` without modification is incorrect. Doing this will need to handle `0` as a special case of doing nothing, 
```suggestion
            0 => {},
            1..=10 => now,
```